homework9/hw1.py

```diff
@@ -0,0 +1,143 @@
+"""
+Write a function that merges integer from sorted files and returns an iterator
+
+file1.txt:
+1
+3
+5
+
+file2.txt:
+2
+4
+6
+
+# >>> list(merge_sorted_files(["file1.txt", "file2.txt"]))
+[1, 2, 3, 4, 5, 6]
+"""
+from contextlib import ExitStack
+from pathlib import Path
+from typing import Iterable, Iterator, List, Union
+
+
+def lines_from_file_generator(file) -> Iterable:
+    yield from (int(line) for line in file)
+
+
+def merge_sorted_files(file_list: List[Union[Path, str]]) -> Iterator:
+    with ExitStack() as stack:
+        files = [stack.enter_context(open(file)) for file in file_list]
+        # Can expand list comprehension into "for loop" with exception FileNotFoundError and ignoring non-existing files
+        # yield from merge_sorted_lists(*map(lambda x: list(map(int, x.read().split("\n"))), files))
+
+        yield from merge_sorted_lists_gen(list(map(lines_from_file_generator, files)))  # More readable and testable
```

Comment:
Now you load all files to memory here. You should not create list at all.

Reply:
I use `list()` here for type conversion from map of generators, to list of generators, but yes I was loading all data into `result` list in `merge_sorted_lists_gen`. Now it is fixed.

Reply:
![image](https://user-images.githubusercontent.com/19683603/114912270-78f2f180-9e28-11eb-9fe9-cbcebfa35929.png)